File: othello_csharp/Othello/Othello.cs
//==========================================================
// Class Othello
// Play Othello (Reversi) on the command line
// https://en.wikipedia.org/wiki/Reversi
// Akseli Lukkarila
// 2019-2023
//==========================================================

using System;
using System.CommandLine;
using System.CommandLine.Invocation;
using Pastel;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.IO;

namespace Othello
{
    public class Program
    {
        /// Gameplay loop and main logic.
        internal class Othello
        {
            public static int MIN_BOARD_SIZE = 4;
            public static int MAX_BOARD_SIZE = 10;
            public static int DEFAULT_BOARD_SIZE = 8;

            private Board _board;
            private Player _playerBlack;
            private Player _playerWhite;
            private int _roundsPlayed;
            private int _gamesPlayed;
            private readonly Settings _settings;
            private List<string> _gameLog = new();

            private Othello(Settings settings)
            {
                _board = new Board(settings.BoardSize);
                _settings = settings;
                _gamesPlayed = 0;
                _playerBlack = Player.Black(settings.ToPlayerSettings());
                _playerWhite = Player.White(settings.ToPlayerSettings());
                _roundsPlayed = 0;
            }

            /// Play one full game of Othello.
            private void Play()
            {
                while (true)
                {
                    InitGame();
                    GameLoop();
                    PrintResult();
                    if (_settings.ShowLog)
                    {
                        PrintLog();
                    }
                    if (_settings.AutoplayMode || !GetAnswer("Would you like to play again"))
                    {
                        break;
                    }
                }
            }

            /// Initialize game board and players for a new game.
            private void InitGame()
            {
                if (_gamesPlayed > 0)
                {
                    _board = new Board(_settings.BoardSize);
                    _playerBlack.Reset();
                    _playerWhite.Reset();
                    _roundsPlayed = 0;
                    _gameLog.Clear();
                }

                if (_settings.AutoplayMode)
                {
                    // Computer plays both
                    _playerWhite.SetHuman(false);
                    _playerBlack.SetHuman(false);
                }
                else if (_settings.QuickStart)
                {
                    // Default: play as black against white computer player
                    _playerWhite.SetHuman(false);
                }
                else if (GetAnswer("Would you like to play against the computer"))
                {
                    if (GetAnswer("Would you like to play as black or white", "b", "w"))
                    {
                        _playerWhite.SetHuman(false);
                    }
                    else
                    {
                        _playerBlack.SetHuman(false);
                    }
                }

                Console.WriteLine("\nPlayers:".Pastel(Color.Silver));
                PrintStatus();
            }

            /// Keep making moves until both players can't make a move any more.
            private void GameLoop()
            {
                while (_board.CanPlay() && (_playerBlack.canPlay || _playerWhite.canPlay))
                {
                    ++_roundsPlayed;
                    Console.WriteLine($"\n=========== ROUND: {_roundsPlayed} ===========");
                    foreach (Player player in new Player[] { _playerBlack, _playerWhite })
                    {
                        var result = player.PlayOneMove(_board);
                        if (result != null)
                        {
                            _gameLog.Add($"{result};{_board.ToLogEntry()}");
                        }
                        Console.WriteLine("--------------------------------");
                    }
                }
                ++_gamesPlayed;
            }

            private void PrintLog()
            {
                string formattedLog = "";
                int index = 1;
                foreach (string line in _gameLog)
                {
                    formattedLog += $"{index++:00}: {line}";
                    if (index <= _gameLog.Count)
                    {
                        formattedLog += "\n";
                    }
                }

                var hexHash = Utils.CalculateSHA256(formattedLog);
                ColorPrint.WriteLine("Game log:", Color.Yellow);
                Console.WriteLine(formattedLog);
                Console.WriteLine(hexHash);
            }

            /// Print ending status and winner info.
            private void PrintResult()
            {
                Console.WriteLine("\n================================");
                ColorPrint.WriteLine("The game is finished!", Color.Green);
                Console.WriteLine("Result:".Pastel(Color.Silver));
                PrintStatus();
                Console.WriteLine("");

                var winner = _board.Result();
                if (winner == Disk.Empty)
                {
                    Console.WriteLine("The game ended in a tie...\n");
                }
                else
                {
                    Console.WriteLine($"The winner is {winner.Name()}!\n");
                }
            }

            /// Print current board and player info.
            private void PrintStatus()
            {
                Console.WriteLine(_playerBlack);
                Console.WriteLine(_playerWhite);
                Console.WriteLine($"\n{_board}");
            }

            /// Ask a question with two options, and return bool from user answer.
            private static bool GetAnswer(string question, string yes = "y", string no = "n")
            {
                Console.Write($"{question} ({yes}/{no})? ");
                var ans = Console.ReadLine();
                return !string.IsNullOrEmpty(ans)
                    && string.Equals(ans, yes, StringComparison.CurrentCultureIgnoreCase);
            }

            /// Ask and return the desired board size.
            private static int GetBoardSize()
            {
                Console.Write($"Choose board size (default is {DEFAULT_BOARD_SIZE}): ");
                if (int.TryParse(Console.ReadLine(), out var boardSize))
                {
                    if (boardSize < Othello.MIN_BOARD_SIZE || boardSize > Othello.MAX_BOARD_SIZE)
                    {
                        ColorPrint.Warn($"Limiting board size to valid range {Othello.MIN_BOARD_SIZE}...{Othello.MAX_BOARD_SIZE}");
                    }
                    return Math.Max(
                        Othello.MIN_BOARD_SIZE,
                        Math.Min(boardSize, Othello.MAX_BOARD_SIZE)
                    );
                }
                ColorPrint.Warn($"Invalid size, defaulting to {Othello.DEFAULT_BOARD_SIZE}...");
                return Othello.DEFAULT_BOARD_SIZE;
            }

            public static int Main(string[] args)
            {
                var size = new Argument<int?>
                    (name: "size",
                    description: "Optional board size",
                    getDefaultValue: () => null);

                var autoplay = new Option<bool>(
                   name: "--autoplay",
                   description: "Enable autoplay mode");
                autoplay.AddAlias("-a");

                var useDefaultSettings = new Option<bool>(
                   name: "--default",
                   description: "Play with default settings");
                useDefaultSettings.AddAlias("-d");

                var showLog = new Option<bool>(
                   name: "--log",
                   description: "Show log after a game");
                showLog.AddAlias("-l");

                var hideHelpers = new Option<bool>(
                   name: "--no-helpers",
                   description: "Hide disk placement hints");
                hideHelpers.AddAlias("-n");

                var testMode = new Option<bool>(
                   name: "--test",
                   description: "Enable test mode");
                testMode.AddAlias("-t");

                var version = new Option<bool>(
                   name: "-v",
                   description: "Print version and exit");

                var rootCommand = new RootCommand("A simple Othello CLI game implementation")
                {
                    size,
                    autoplay,
                    useDefaultSettings,
                    showLog,
                    hideHelpers,
                    testMode,
                    version
                };

                ColorPrint.WriteLine("OTHELLO GAME - C#", Color.Green);

                rootCommand.SetHandler((size, autoplay, useDefaultSettings, showLog, hideHelpers, testMode, version) =>
                    {
                        if (version)
                        {
                            Console.WriteLine($"Othello C# {Utils.VersionInfo()}");
                            Environment.Exit(0);
                        }

                        int boardSize;
                        if (size != null)
                        {
                            boardSize = size.Value;
                            if (boardSize < Othello.MIN_BOARD_SIZE || boardSize > Othello.MAX_BOARD_SIZE)
                            {
                                ColorPrint.Error($"Unsupported board size: {boardSize}");
                                Environment.Exit(1);
                            }
                            Console.WriteLine($"Using board size: {boardSize}");

                        }
                        else if (autoplay || useDefaultSettings)
                        {
                            boardSize = DEFAULT_BOARD_SIZE;
                        }
                        else
                        {
                            // Ask user for board size
                            boardSize = GetBoardSize();
                        }

                        Settings settings = new(boardSize, autoplay, useDefaultSettings, showLog, !hideHelpers, testMode);

                        var game = new Othello(settings);
                        game.Play();
                    },
                    size, autoplay, useDefaultSettings, showLog, hideHelpers, testMode, version);

                try
                {
                    return rootCommand.Invoke(args);
                }
                catch (OperationCanceledException)
                {
                    Console.WriteLine("\naborted...");
                    return 1;
                }
                catch (Exception ex)
                {
                    ColorPrint.Error($"An exception occurred: {ex.Message}");
                    return 1;
                }
            }
        }
    }
}

File: swift/Sources/othello_swift/othello.swift
//==========================================================
// Othello
// Play Othello (Reversi) on the command line
// https://en.wikipedia.org/wiki/Reversi
// Akseli Lukkarila
// 2019-2024
//==========================================================

import ColorizeSwift

/// Gameplay loop and main logic.
class Othello {
    var board: Board
    var settings: Settings
    var playerBlack: Player
    var playerWhite: Player
    var roundsPlayed: Int = 0
    var gamesPlayed: Int = 0
    var gameLog: [String] = []

    init(_ settings: Settings) {
        self.board = Board(size: settings.boardSize)
        self.settings = settings
        self.playerBlack = Player.black(settings.toPlayerSettings())
        self.playerWhite = Player.white(settings.toPlayerSettings())
    }

    /// Play one full game of Othello.
    func play() {
        while true {
            self.initGame()
            self.gameLoop()
            self.printResult()
            if self.settings.showLog {
                self.printLog()
            }
            if self.settings.autoplayMode || !self.getAnswer("Would you like to play again") {
                break
            }
        }
    }

    /// Initialize game board and players for a new game.
    func initGame() {
        if self.gamesPlayed > 0 {
            self.board = Board(size: self.settings.boardSize)
            self.playerBlack.reset()
            self.playerWhite.reset()
            self.roundsPlayed = 0
        }

        if self.settings.autoplayMode {
            // Computer plays both
            self.playerWhite.setHuman(false)
            self.playerBlack.setHuman(false)
        } else if self.settings.useDefaults {
            // Default: play as black against white computer player
            self.playerWhite.setHuman(false)
        } else if self.getAnswer("Would you like to play against the computer") {
            if self.getAnswer("Would you like to play as black or white", yes: "b", no: "w") {
                self.playerWhite.setHuman(false)
            } else {
                self.playerBlack.setHuman(false)
            }
        }

        print("\nPlayers:".bold())
        self.printStatus()
    }

    /// Keep making moves until both players can't make a move any more.
    func gameLoop() {
        while self.board.canPlay() && (self.playerBlack.canPlay || self.playerWhite.canPlay) {
            self.roundsPlayed += 1
            printBold("\n=========== ROUND: \(self.roundsPlayed) ===========")
            for player in [self.playerBlack, self.playerWhite] {
                if let result = player.playOneMove(board: &self.board) {
                    self.gameLog.append("\(result);\(self.board.toLogEntry())")
                }
            }
            print("-------------------------------")
        }
    }

    /// Print game log which shows all moves made and the game board state after each move.
    func printLog() {
        var formattedLog = ""
        var index = 1
        for line in self.gameLog {
            formattedLog += String(format: "%02d: %@", index, line)
            if index < self.gameLog.count {
                formattedLog += "\n"
            }
            index += 1
        }

        let hexHash = calculateSHA256(formattedLog)

        printBold("Game log:", color: TerminalColor.yellow)
        print(formattedLog)
        print(hexHash)
    }

    /// Print ending status and winner info.
    func printResult() {
        print("\n===============================".bold())
        print("The game is finished!\n".green())
        print("Result:".bold())
        self.printStatus()
        print("")

        let winner = self.board.result()
        switch winner {
            case Disk.empty:
                print("The game ended in a tie...\n")
            default:
                print("The winner is \(winner)!\n")
        }
    }

    /// Print current board and player info.
    func printStatus() {
        print(self.playerBlack)
        print(self.playerWhite, terminator: "\n\n")
        print(self.board)
    }

    /// Ask a question with two options, and return bool from user answer.
    func getAnswer(_ question: String, yes: String = "y", no: String = "n") -> Bool {
        print("\(question) (\(yes)/\(no))? ", terminator: "")
        let ans = readLine()
        return ans?.lowercased() == yes.lowercased()
    }

    /// Ask and return the desired board size.
    class func getBoardSize() -> Int {
        while true {
            print("Choose board size (default is \(DEFAULT_BOARD_SIZE)): ", terminator: "")
            if let input = readLine() {
                // TODO: create clamp method
                if let size = Int(input) {
                    return max(MIN_BOARD_SIZE, min(size, MAX_BOARD_SIZE))
                }
            }
            printError("give a valid number...")
        }
    }
}
